### PR TITLE
feat: stats recalculate button with diff stats backfill

### PR DESCRIPTION
**Pull request by DGouron**
## Summary

- **Recalculate button** in dashboard stats section header triggers stats recomputation
- **Backfill mechanism** fetches missing `diffStats` for old reviews via GitLab/GitHub APIs
- Processes in **batches of 10** with 2-second delay between batches to respect rate limits
- **Real-time progress** via WebSocket (`backfill-progress` / `backfill-complete` messages)
- Dashboard shows progress badge "N/total reviews" during processing
- Failed individual fetches set `diffStats: null` and continue (never blocks)

## Architecture

- `BackfillDiffStats` use case: orchestrates batch fetching with progress callback
- `RecalculateProjectStats` use case: recomputes all aggregate fields from raw reviews
- `POST /api/stats/recalculate` endpoint: fire-and-forget, returns immediately
- WebSocket broadcast for live progress updates to dashboard

## Test plan

- [x] `yarn verify` passes (1080 tests, 0 errors)
- [x] Manual: Recalculate button triggers backfill with progress indicator
- [x] Manual: Stats refresh after completion
- [ ] Manual: Test with GitHub project (only tested with GitLab)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/interface-adapters/controllers/http/stats.routes.ts [hunk 1]
```diff
@@ -49,4 +63,65 @@ export const statsRoutes: FastifyPluginAsync<StatsRoutesOptions> = async (
...
+    );
+
+    if (!repository) {
+      reply.status(404).send({ error: 'Projet non trouvé dans la configuration' });
+      return;
+    }
+
+    const { diffStatsFetchGateways, broadcastBackfillProgress, logger } = options;
+
+    const executeRecalculation = async () => {
```
Comment (DGouron) [inline src/interface-adapters/controllers/http/stats.routes.ts]: **[BLOCKING] Business logic in controller — violates Humble Object pattern**

`stats.routes.ts:89-121`

The `executeRecalculation` function contains business orchestration: platform selection (line 92), conditional backfill invocation, use case coordination, and progress broadcasting. Controllers must be "humble" — too simple to need tests.

> "Make your controllers so simple that they are not worth testing." — Robert C. Martin, Clean Architecture, Chapter 23

**Fix**: Extract this into a dedicated use case (e.g., `recalculateWithBackfill.usecase.ts`) that takes `projectPath`, `shouldBackfill`, `platform` and orchestrates `backfillDiffStats` + `recalculateProjectStats`. The controller should only parse the request, call the use case, and return the response.

File: src/interface-adapters/controllers/http/stats.routes.ts [hunk 2]
```diff
@@ -49,4 +63,65 @@ export const statsRoutes: FastifyPluginAsync<StatsRoutesOptions> = async (
 
     return { projects: allStats };
   });
+
+  fastify.post('/api/stats/recalculate', async (request, reply) => {
+    const body = request.body as { path?: string; backfill?: boolean } | null;
```
Comment (DGouron) [inline src/interface-adapters/controllers/http/stats.routes.ts]: **[IMPORTANT] `as` type assertion — use Zod guard at boundary**

`stats.routes.ts:68`

```typescript
const body = request.body as { path?: string; backfill?: boolean } | null;
```

Type assertions are forbidden per CLAUDE.md. This is a system boundary (HTTP input) — exactly where Zod validation should be used.

**Fix**: Create a Zod schema and use `safeParse`:
```typescript
const recalculateBodySchema = z.object({
  path: z.string().min(1),
  backfill: z.boolean().optional().default(false),
});
const parsed = recalculateBodySchema.safeParse(request.body);
if (!parsed.success) {
  reply.status(400).send({ error: 'Invalid request body' });
  return;
}
```

File: src/interface-adapters/gateways/diffStatsFetch.github.gateway.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import type { DiffStatsFetchGateway } from '@/entities/diffStats/diffStatsFetch.gateway.js';
+import type { DiffStats } from '@/entities/diffStats/diffStats.js';
+
+export type CommandExecutor = (command: string) => string;
```
Comment (DGouron) [inline src/interface-adapters/gateways/diffStatsFetch.github.gateway.ts]: **[IMPORTANT] Duplicated `CommandExecutor` type**

`diffStatsFetch.github.gateway.ts:4` and `diffStatsFetch.gitlab.gateway.ts:4`

```typescript
export type CommandExecutor = (command: string) => string;
```

This exact type is defined in both gateway files. DRY violation with 100% duplication across 2 files.

**Fix**: Define `CommandExecutor` once in a shared location (e.g., `src/shared/foundation/commandExecutor.ts` or alongside the existing executor patterns in the codebase).

File: src/usecases/stats/recalculateProjectStats.usecase.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
+import type { StatsGateway } from '@/interface-adapters/gateways/stats.gateway.js';
+import type { ProjectStats } from '@/services/statsService.js';
```
Comment (DGouron) [inline src/usecases/stats/recalculateProjectStats.usecase.ts]: **[IMPORTANT] Use case imports domain type from services layer**

`recalculateProjectStats.usecase.ts:2`

```typescript
import type { ProjectStats } from '@/services/statsService.js';
```

`ProjectStats` is a domain type but lives in `src/services/statsService.ts`, which mixes type definitions with file I/O operations (`readFileSync`, `writeFileSync`). The use case should not depend on this infrastructure file.

**Fix**: Extract `ProjectStats` and `ReviewStats` types to `src/entities/stats/projectStats.ts`. The service file should import from entities, not the other way around.

**Comment by DGouron** [comment]
# Code Review - PR #124 (feat: stats recalculate button with diff stats backfill)

**Date**: 2026-03-15
**Reviewer**: Claude Code (Mentor Mode)
**Branch**: `feat/stats-recalculate-backfill`
**Modified files**: 27 (+1940/-31 lines)

---

## Executive Summary

| Audit | Score | Verdict |
|-------|-------|---------|
| **Clean Architecture** | 5/10 | Business logic in controller, Dependency Rule violations |
| **Strategic DDD** | 7/10 | Good context isolation, domain types misplaced in services |
| **TypeScript Best Practices** | 6/10 | `as` assertions at boundaries, missing Zod guards |
| **SOLID** | 7/10 | SRP and DIP violations in controller and use case imports |
| **Testing** | 7/10 | Strong use case coverage, trivial entity tests, fragile async test |
| **Security** | 7/10 | No input validation at boundary, low-risk command injection pattern |

**Overall Score: 6.5/10** - Solid feature implementation with good test coverage, but architectural violations need to be fixed before merge.

---

## Blocking Corrections (before merge)

### 1. Business logic in controller

**Files**: `src/interface-adapters/controllers/http/stats.routes.ts:89-121`

The `executeRecalculation` inner function contains business orchestration logic: platform selection, conditional backfill invocation, use case coordination, and progress broadcasting. Controllers must be "humble" — too simple to need tests.

> "Make your controllers so simple that they are not worth testing."
> — Robert C. Martin, Clean Architecture, Chapter 23

**Fix**: Extract into a dedicated use case (e.g., `recalculateWithBackfill.usecase.ts`) that takes `projectPath`, `shouldBackfill`, `platform` and orchestrates `backfillDiffStats` + `recalculateProjectStats`. The controller should only parse the request, call the use case, and return the response.

---

### 2. Dependency Rule violation — use cases import from outer layers

**Files**: `src/usecases/stats/backfillDiffStats.usecase.ts:1`, `src/usecases/stats/recalculateProjectStats.usecase.ts:1-2`

Use cases import `StatsGateway` from `@/interface-adapters/gateways/` and `ProjectStats` from `@/services/`. The Dependency Rule requires all source code dependencies to point inward.

> "Source code dependencies must point only inward, toward higher-level policies."
> — Robert C. Martin, Clean Architecture, Chapter 22

**Fix**: Move `StatsGateway` interface to `src/entities/stats/stats.gateway.ts`. Move `ProjectStats`/`ReviewStats` types to `src/entities/stats/projectStats.ts`.

---

## Important Corrections (this week)

### 3. `as` type assertion at HTTP boundary

**File**: `src/interface-adapters/controllers/http/stats.routes.ts:68`

```typescript
const body = request.body as { path?: string; backfill?: boolean } | null;
```

Type assertions are forbidden per project rules. Use a Zod guard at this system boundary.

---

### 4. Duplicated `CommandExecutor` type

**Files**: `diffStatsFetch.github.gateway.ts:4`, `diffStatsFetch.gitlab.gateway.ts:4`

Both files define identical `export type CommandExecutor = (command: string) => string;`. Define once in a shared location.

---

### 5. JSDoc removed from public API types

**File**: `src/services/statsService.ts`

JSDoc comments on `ReviewStats` and `ProjectStats` were removed. Per project rules, JSDoc for public APIs should be preserved.

---

## Improvements (backlog)

| # | Issue | File |
|---|-------|------|
| 6 | Trivial entity type tests (test TypeScript type system, not behavior) | `backfillProgress.test.ts`, `diffStats.test.ts` |
| 7 | `statsCharts.js` pure functions (`getIsoWeek`, data aggregation) could be extracted and tested | `statsCharts.js` |
| 8 | `diffStats?: DiffStats \| null` uses `undefined` via `?` — contradicts "undefined banned" rule | `statsService.ts` |

---

## Positive Observations

| Aspect | Observation |
|--------|-------------|
| Gateway pattern | `DiffStatsFetchGateway` contract correctly in entities, implementations in interface-adapters |
| Use case design | `backfillDiffStats` and `recalculateProjectStats` are well-scoped, single-responsibility |
| Test coverage | 10 test files for 8 production files, good edge case coverage |
| Batch processing | Rate-limiting with configurable batch size and delay shows production awareness |
| Error resilience | Individual fetch failures set `diffStats: null` and continue — never blocks |
| Import cleanup | Fixes existing relative imports to use `@/` alias |

---

## Pre-Merge Checklist

- [ ] Extract business orchestration from stats.routes.ts into a use case
- [ ] Move StatsGateway interface to entities layer
- [ ] Move ProjectStats/ReviewStats types to entities layer
- [ ] Replace `as` assertion with Zod guard for recalculate endpoint body
- [ ] Deduplicate CommandExecutor type
- [ ] `yarn verify` passes without error

---

## Recommended Skills for Corrections

1. **Business logic in controller** -> `/architecture` to design the orchestrating use case
2. **Dependency Rule violation** -> `/solid` to understand DIP and dependency direction
3. **Missing Zod guard** -> `/tdd` to create the guard with RED-GREEN-REFACTOR cycle
4. **Before deduplicating CommandExecutor** -> `/anti-overengineering` to validate approach

---

[REVIEW_STATS:blocking=2:warnings=3:suggestions=3:score=6.5]
